## Phase 2 (Data Understanding): Statistische Analyse

* Autorin: Anna (i3-Versicherung)
* Webseite: [Data Science Training - Kapitel 2](https://data-science.training/kapitel-2/)
* Datum: 23.03.2023

Wir führen die Trainings- und Testdaten zusammen, sehen uns die Daten an und berechnen einige statistische Größen.

In [4]:
# Pandas Paket (Package) importieren
#  Datenstrukturen und Datenanalyse, I/O
#  https://pandas.pydata.org/pandas-docs/stable/
import pandas as pd

In [5]:
# Trainings- und Testdaten als Pandas Data Frame (df) aus CSV-Dateien laden
#  (KNIME: "CSV Reader")
df_train = pd.read_csv('../../data/titanic/original/train.csv')
df_test  = pd.read_csv('../../data/titanic/original/test.csv')

In [6]:
# Daten zusammenführen
#  (KNIME "Concatenate")
# Hierzu benutzen wir die Funktion concat(),
# wobei wir eine Liste mit den einzelnen Data Frames als Argument übergeben.
df = pd.concat([df_train, df_test], ignore_index=True)

In [7]:
# Statistische Informationen
#  (KNIME: "Statistics")
df.describe()

PassengerId    Survived       Pclass          Age        SibSp  \
count  1309.000000  891.000000  1309.000000  1046.000000  1309.000000   
mean    655.000000    0.383838     2.294882    29.881138     0.498854   
std     378.020061    0.486592     0.837836    14.413493     1.041658   
min       1.000000    0.000000     1.000000     0.170000     0.000000   
25%     328.000000    0.000000     2.000000    21.000000     0.000000   
50%     655.000000    0.000000     3.000000    28.000000     0.000000   
75%     982.000000    1.000000     3.000000    39.000000     1.000000   
max    1309.000000    1.000000     3.000000    80.000000     8.000000   

             Parch         Fare  
count  1309.000000  1308.000000  
mean      0.385027    33.295479  
std       0.865560    51.758668  
min       0.000000     0.000000  
25%       0.000000     7.895800  
50%       0.000000    14.454200  
75%       0.000000    31.275000  
max       9.000000   512.329200

In [8]:
# Fehlende Werte
#  (KNIME: "Statistics")
df.isnull().sum()

PassengerId       0
Survived        418
Pclass            0
Name              0
Sex               0
Age             263
SibSp             0
Parch             0
Ticket            0
Fare              1
Cabin          1014
Embarked          2
dtype: int64

In [9]:
# Datentypen bestmöglich automatisch konvertieren: z.B. object => string
df = df.convert_dtypes()
# Ausgabe
print(df.dtypes)

PassengerId             Int64
Survived                Int64
Pclass                  Int64
Name           string[python]
Sex            string[python]
Age                   Float64
SibSp                   Int64
Parch                   Int64
Ticket         string[python]
Fare                  Float64
Cabin          string[python]
Embarked       string[python]
dtype: object


In [10]:
# Überblick: Datentypen und Anzahl nicht-fehlender Daten anzeigen
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  1309 non-null   Int64  
 1   Survived     891 non-null    Int64  
 2   Pclass       1309 non-null   Int64  
 3   Name         1309 non-null   string 
 4   Sex          1309 non-null   string 
 5   Age          1046 non-null   Float64
 6   SibSp        1309 non-null   Int64  
 7   Parch        1309 non-null   Int64  
 8   Ticket       1309 non-null   string 
 9   Fare         1308 non-null   Float64
 10  Cabin        295 non-null    string 
 11  Embarked     1307 non-null   string 
dtypes: Float64(2), Int64(5), string(5)
memory usage: 131.8 KB


In [11]:
# Die ersten und letzten Datensätze als formatierte Tabelle anzeigen
display(df)

PassengerId  Survived  Pclass  \
0               1         0       3   
1               2         1       1   
2               3         1       3   
3               4         1       1   
4               5         0       3   
...           ...       ...     ...   
1304         1305      <NA>       3   
1305         1306      <NA>       1   
1306         1307      <NA>       3   
1307         1308      <NA>       3   
1308         1309      <NA>       3   

                                                   Name     Sex   Age  SibSp  \
0                               Braund, Mr. Owen Harris    male  22.0      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss. Laina  female  26.0      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                              Allen, Mr. William Henry    male  35.0      0   
...                                                 ...     ...   ...    ...   
1304                                 Spector, Mr. Woolf    male  <NA>      0   
1305                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
1306                       Saether, Mr. Simon Sivertsen    male  38.5      0   
1307                                Ware, Mr. Frederick    male  <NA>      0   
1308                           Peter, Master. Michael J    male  <NA>      1   

      Parch              Ticket     Fare Cabin Embarked  
0         0           A/5 21171     7.25  <NA>        S  
1         0            PC 17599  71.2833   C85        C  
2         0    STON/O2. 3101282    7.925  <NA>        S  
3         0              113803     53.1  C123        S  
4         0              373450     8.05  <NA>        S  
...     ...                 ...      ...   ...      ...  
1304      0           A.5. 3236     8.05  <NA>        S  
1305      0            PC 17758    108.9  C105        C  
1306      0  SOTON/O.Q. 3101262     7.25  <NA>        S  
1307      0              359309     8.05  <NA>        S  
1308      1                2668  22.3583  <NA>        C  

[1309 rows x 12 columns]

### Ergebnisse der Datensichtung und der statistischen Analysen

Beschreibungen (siehe auch Kaggle: https://www.kaggle.com/c/titanic/data):

| Attribut    | Beschreibung                       | Domäne         | Skalenniveau | Datentyp | Fehlende Werte |
| ------------| -----------------------------------| ---------------| -------------| ---------| ---------------|
| PassengerId | Eindeutige Nummer                  | 1 bis 1309     | nominal      | Int      |                |
| Survived    | Überlebt                           | 0 oder 1       | nominal      | Int      | Testdaten: 418 |
| Pclass      | Passagierklasse                    | 1, 2, 3        | ordinal      | Int      |                |
| Name        | Passagiername                      | Text           | nominal      | String   |                |
| Sex         | Geschlecht                         | female, male   | nominal      | String   |                |
| Age         | Alter in Jahren                    | 0,17 bis 80    | kardinal     | Float    | 263            |
| SibSp       | # Geschwister & Ehepartner an Bord | 0 bis 8        | kardinal     | Int      |                |
| Parch       | # Eltern & Kinder an Bord          | 0 bis 9        | kardinal     | Int      |                |
| Ticket      | Ticketnummer                       | alphanumerisch | nominal      | String   |                |
| Fare        | Ticket-Preis in Pfund              | 0 bis 512,3292 | kardinal     | Float    | 1              |
| Cabin       | Kabinennummer                      | alphanumerisch | nominal      | String   | 1014           |
| Embarked    | Abfahrtshafen                      | C, Q, S        | nominal      | String   | 2              |

Legende mit Bedeutungen der Werte:
* Survived: 0 = Nein, 1 = Ja
* Pclass: 1 = erste, 2 = zweite, 3 = dritte 
* Embarked: C = Cherbourg, Q = Queenstown, S = Southampton

Erste Erkenntnisse:

* Die PassengerId ist ein Schlüsselattribut zur eindeutigen Identifizierung eines Passagiers. Es enthält aber keine Informationen, die für die Datenanalyse relevant sind.

* In den Namen stecken weitere Informationen wie z.B. die Anrede, der Titel, der Mädchenname (bei Frauen) usw.

* Das Alter wurde manchmal nur geschätzt (Kommazahl) und fehlt auch sehr häufig.

* Die genaue Bedeutung der Ticketnummer ist unklar.

* Der Ticket-Preis bezieht sich ggf. auf mehrere Passagiere, die unter der gleichen Ticketnummer gereist sind.

* Die Kabinennummer, falls vorhanden, beginnt mit dem Deck. In der 3. Klasse haben sich vermutlich viele Passagiere eine gemeinsame Kabine geteilt. Diese Daten sind kaum erfasst.